core/drivers/zynqmp_pm.c

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2021 Foundries.io Ltd
+ */
+
+#include <arm.h>
+#include <drivers/zynqmp_pm.h>
+#include <kernel/cache_helpers.h>
+#include <kernel/thread.h>
+#include <mm/core_memprot.h>
+#include <string.h>
+#include <tee/cache.h>
+#include <tee_api_types.h>
+#include <types_ext.h>
+#include <utee_defines.h>
+
+#define IS_ALIGNED(x, a) (((x) & ((typeof(x))(a) - 1)) == 0)
+
+/*
+ * For additional details about ZynqMP specific SMC ID's and PM request
+ * handling in TF-A check
+ * https://xilinx-wiki.atlassian.net/wiki/spaces/A/pages/18842107/Arm+Trusted+Firmware
+ */
+#define EFUSE_ACCESS_SMC	0xC2000035
+#define EFUSE_NOT_ENABLED	29
+
+enum efuse_op {EFUSE_READ = 0, EFUSE_WRITE = 1};
+
+/*
+ * Stores all required details to
+ * read/write efuse memory.
+ * @src:        paddr_t of the buffer to store the data to be write/read
+ *              buffer should be cache aligned
+ * @size:       number of 32-bit words to be read/written
+ * @offset:     offset to be read/written
+ * @flag:       EFUSE_READ  - represents efuse read operation
+ *              EFUSE_WRITE - represents efuse write operation
+ * @pufuserfuse:0 - represents non-puf efuses, offset is used for read/write
+ *              1 - represents puf user fuse row number.
+ */
+struct xilinx_efuse {
+	uint64_t src;
+	uint32_t size;
+	uint32_t offset;
+	uint32_t flag;
+	uint32_t pufuserfuse;
+};
+
+static uint32_t zynqmp_sip_call(uint32_t pm_api_id, uint32_t arg0,
+				uint32_t arg1, uint32_t arg2, uint32_t arg3,
+				uint32_t *payload)
+{
+	struct thread_smc_args args = {
+		.a0 = pm_api_id,
+		.a1 = reg_pair_to_64(arg1, arg0),
+		.a2 = reg_pair_to_64(arg3, arg2),
+	};
+
+	thread_smccc(&args);
+
+	if (payload)
+		*payload = args.a0 >> 32;
+
+	return args.a0;
+}
+
+static void *alloc_aligned_zeroed(size_t *sz)
+{
+	size_t alloc_size = 0;
+	size_t cacheline_size = 0;
+	void *ptr = NULL;
+
+	if (!sz)
```

Comment:
`if (!sz)` or `if (!sz || !*sz)` ?

Reply:
only: `if (!sz)`, in case `*sz == 0`, payload then will be filled with zeros in the next line, so no problem here